src/Dependencies/Collections/ImmutableSegmentedHashSet`1.cs

```diff
@@ -12,148 +12,277 @@ namespace Microsoft.CodeAnalysis.Collections
     internal readonly partial struct ImmutableSegmentedHashSet<T> : IImmutableSet<T>, ISet<T>, ICollection, IEquatable<ImmutableSegmentedHashSet<T>>
     {
         /// <inheritdoc cref="ImmutableHashSet{T}.Empty"/>
-        public static readonly ImmutableSegmentedHashSet<T> Empty;
+        public static readonly ImmutableSegmentedHashSet<T> Empty = new(new SegmentedHashSet<T>());
+
+        private readonly SegmentedHashSet<T> _set;
+
+        private ImmutableSegmentedHashSet(SegmentedHashSet<T> set)
+        {
+            _set = set;
+        }
 
         /// <inheritdoc cref="ImmutableHashSet{T}.KeyComparer"/>
-        public IEqualityComparer<T> KeyComparer => throw null!;
+        public IEqualityComparer<T> KeyComparer => _set.Comparer;
 
         /// <inheritdoc cref="ImmutableHashSet{T}.Count"/>
-        public int Count => throw null!;
+        public int Count => _set.Count;
+
+        public bool IsDefault => _set == null;
 
         /// <inheritdoc cref="ImmutableHashSet{T}.IsEmpty"/>
-        public bool IsEmpty => throw null!;
+        public bool IsEmpty => _set.Count == 0;
 
-        bool ICollection<T>.IsReadOnly => throw null!;
+        bool ICollection<T>.IsReadOnly => true;
 
-        bool ICollection.IsSynchronized => throw null!;
+        bool ICollection.IsSynchronized => true;
 
-        object ICollection.SyncRoot => throw null!;
+        object ICollection.SyncRoot => _set;
 
         public static bool operator ==(ImmutableSegmentedHashSet<T> left, ImmutableSegmentedHashSet<T> right)
-            => throw null!;
+            => left.Equals(right);
 
         public static bool operator !=(ImmutableSegmentedHashSet<T> left, ImmutableSegmentedHashSet<T> right)
-            => throw null!;
+            => !left.Equals(right);
 
         public static bool operator ==(ImmutableSegmentedHashSet<T>? left, ImmutableSegmentedHashSet<T>? right)
-            => throw null!;
+            => left.GetValueOrDefault().Equals(right.GetValueOrDefault());
 
         public static bool operator !=(ImmutableSegmentedHashSet<T>? left, ImmutableSegmentedHashSet<T>? right)
-            => throw null!;
+            => !left.GetValueOrDefault().Equals(right.GetValueOrDefault());
 
         /// <inheritdoc cref="ImmutableHashSet{T}.Add(T)"/>
         public ImmutableSegmentedHashSet<T> Add(T value)
-            => throw null!;
+        {
+            var self = this;
+
+            if (self.IsEmpty)
+            {
+                var set = new SegmentedHashSet<T>(self.KeyComparer) { value };
+                return new ImmutableSegmentedHashSet<T>(set);
+            }
+            else if (self.Contains(value))
+            {
+                return self;
+            }
+            else
+            {
+                // TODO: Avoid the builder allocation
```

Comment:
Are these todo's from the original source?

Reply:
No, these are ways the current implementation could theoretically be improved. My hope is that if someone sees allocations of certain types in the future, they are directed to the most straightforward way to eliminate those allocations.

Reply:
Here's an example of how I broke out a `ValueBuilder` nested type to reduce builder allocations in `ImmutableSegmentedList<T>`:
https://github.com/dotnet/roslyn/blob/main/src/Dependencies/Collections/ImmutableSegmentedList%601%2BValueBuilder.cs